ssg/templates.py

```diff
@@ -285,25 +241,49 @@ def get_lang_for_rule(self, rule_id, rule_title, template, language):
         # definition ID so that the build system matches the generated
```

Comment:
These variables, `rule_id` and `rule_title` are hardcoded in tons of templates and macros, and templates are expecting them even if they are rendered for "extra ovals" or even platforms. And it kinda makes sense to not have it named `check_id` or `check_title` because this base value is used not only in checks but in remediations as well (and would be in platforms).

Passing `rule_id` into functions that supposed to be agnostic to this level of abstractions is *very ugly*, but it makes perfect sense in practice: whenever a dev screws a build by changing something in a rule it won't help knowing what template caused it. Other way around is pretty straightforward — every rule has just one template.

Also, platforms would add another slice of complexity here because of second layer of indirection.

The whole template builder code (and rendering) is written upside-down, and IMHO should be rewritten (including Jinja part), but it is a huge scope creep.

One thing I could promise to you right now though: there would be one less extra_env=None in the next PR.

Reply:
Let's forget about introduction of `check_id`, or about renaming `rule_id` to `id` and so on - we both agree that those things are outside of the scope of this PR. However, if something is named `rule_id` and I have to pass it an ID that isn't a rule ID to get what I need, then the code should be refactored, because it confuses its users.
Let's also forget about "extra_env"'s name. On that level of abstraction, it is "stuff for the template that the method should to pass on, and perhaps add some extra stuff there", but we rename it later.

Usage of `rule_id` in the code that returns the template name is slightly unfortunate, because throwing an exception would allow to provide the context on a higher level, but I also don't suggest to change this, as this is not the code touched by this PR.

I suggest to introduce this generic function that would sit between `build_rule`/`build_extra_ovals` and `build_lang` - the `render_lang_file` that you propose is too low-level to be conveniently used directly. In order to do that easily, I propose to move rule ID and rule title to that "extra env" dictionary. You will still need to put something for `get_template_name`, but "extra OVAL ID ..." will do just right for the time being.

Reply:
Okay, I'm looking at a newer code already. So there is a bit of disconnect. In the future that you're not seeing yet there is no `build_lang`: `build_rule` -> `get_lang_contents` -> `render_lang_file`. That's it. The function `get_lang_contents` is the mediator (or would be). But I really don't want to bring another couple dozen lines in this PR. We will have an opportunity to discuss it once more with platform templates rendering present in the code.

Reply:
Great, yes, you are correct, you will have to repeat it, but that's fine. You compose "the stuff for the template", and it is expected that you build that with slightly different intentions in each of those cases. It may look weird, as the "env yaml" keys will still be rule-centric, but we can discuss that later. Right now, it is important that the code allows different way of filling that dictionary.

Regarding `get_lang_contents` - I suggest slightly changing its API to something like
```
get_lang_contents(self, rule_id_context, template, language, extra_env)
```
and remove statements that populate `extra_env` in lower-level functions.

`build_sce.checks` is an interesting one - you actually have the "stuff for the template" at your disposal it just waits to be passed on, so you can even make the somehow simpler: https://github.com/ComplianceAsCode/content/blob/master/ssg/build_sce.py#L128

Reply:
This is a refactoring PR, and it is not done, because the code that has been touched has a significant design smell. Therefore, I won't merge this PR until the smell is dealt with, we don't use future as an argument when reviewing PRs.

Reply:
Python is not a strongly typed language, but there is a reason we don't just pass around all arguments in a generic dictionary.

Reply:
If you insist that implicit required arguments for templates named `rule_id` and `rule_title` is an important problem, and it should be dealt with right away — I shall try and do that.

Reply:
OK So do I understand it well that there won't be any further changes in this PR?

Reply:
I'd prefer the first option. Using a dictionary for passing parameters could make the function less readable because people need to make more effort to discover the actual contents of the dictionary.

Reply:
That dictionary is a list of substitutions, its "customer" is a template, not lower-level functions down the road. So having stuff there doesn't affect the readability, because you don't have reasons to inspect it, it is just a package of static data.

Concerning any contract-related stuff, the only proper way of handling this is to make Jinja more explicit about undefined values, which BTW can be done. If we go for something weaker s.a. assuring that those keys are available to the templates, a more solid way than supplying arguments to higher-level functions would be to check the dictionary further down for keys, e.g. in `render_lang_file`. So I don't see a point here either.

My objection related to the current PR isn't template field names, because those have already been here before the PR. The problem is that the refactored code has confusing design, which can be demonstrated in the relationship between `build_extra_ovals` and `build_lang`. It may look like a small thing, but it isn't, because it creates an implicit assumption, it makes the code confusing, and we refactor exactly in order to avoid such things.